src/backend/dx12/src/root_constants.rs

```diff
@@ -0,0 +1,302 @@
+//! Processing push constant ranges
+//!
+//! This module provides utitlity functions to make push constants root signature
+//! compatible. Root constants are non-overlapping, therefore, the push constant
+//! ranges passed at pipeline layout creation need to be `split` into disjunct
+//! ranges. The disjunct ranges can be then converted into root signature entries.
+
+use hal::pso;
+use std::cmp::Ordering;
+use std::ops::Range;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct RootConstant {
+    pub stages: pso::ShaderStageFlags,
+    pub range: Range<u32>,
+}
+
+impl RootConstant {
+    fn is_empty(&self) -> bool {
+        self.range.end <= self.range.start
+    }
+
+    // Divide a root constant into two separate ranges depending on the overlap
+    // with another root constant.
+    fn divide(self, other: &RootConstant) -> (RootConstant, RootConstant) {
+        let left = RootConstant {
+            stages: self.stages,
+            range: self.range.start..other.range.start,
+        };
+
+        let right = RootConstant {
+            stages: self.stages,
+            range: other.range.start..self.range.end,
+        };
+
+        (left, right)
+    }
+}
+
+impl PartialOrd for RootConstant {
+    fn partial_cmp(&self, other: &RootConstant) -> Option<Ordering> {
+        Some(self.range.start.cmp(&other.range.start)
```

Comment:
technically if we reorder `stages` to go last, we could derive the `Ord` and `PartialOrd`, right?

Reply:
Not sure what you mean.

Reply:
Nvm, I see that `Range` doesn't implement `PartialOrd`. Was thinking about putting `#[derive()]` instead of manual implementation, but it's not possible.